lib/extract-templates.js

```diff
@@ -0,0 +1,133 @@
+
+export const SUPPORTED_EXTENSIONS = ['.js', '.ts', '.gjs', '.gts'];
+const LOCATION_START = Object.freeze({ line: 0, column: 0 });
+/**
+ * @typedef {object} TemplateInfo
+ * @property {number} line
+ * @property {number} column
+ * @property {string} template
+ * @property {boolean} isEmbedded
+ *
+ * @param {string} moduleSource
+ * @param {string} relativePath
+ *
+ * @returns {TemplateInfo[]}
+ */
+export function extractTemplates(moduleSource, relativePath) {
+  // If no relativePath is present, assuming we might have templates.
+  let mightHaveTemplates = relativePath
+    ? SUPPORTED_EXTENSIONS.some((ext) => relativePath.endsWith(ext))
+    : true;
+
+  if (!mightHaveTemplates) {
+    return [makeTemplateInfo(LOCATION_START, moduleSource)];
+  }
+
+  let parsed = parseTemplates(moduleSource, relativePath, 'template');
+
+  // If after parsing we found no templates and we had no relative path,
+  // then we assume we had a hbs file in input.
+  if (parsed.length === 0 && !relativePath) {
+    return [makeTemplateInfo(LOCATION_START, moduleSource)];
+  }
+
+  let result = parsed.map((templateInfo) => {
+    let { start, end } = templateInfo;
+    let templateStart = start.index + start[0].length;
+
+    let template = moduleSource.slice(templateStart, end.index);
+
+    return makeTemplateInfo(coordinatesOf(moduleSource, templateStart, end.index), template, true);
+  });
+
+  return result;
+}
+
+/**
+ * @param {object} location
+ * @param {number} location.line
+ * @param {number} location.column
+ * @param {string} template
+ * @param {boolean} isEmbedded
+ *
+ * @returns {TemplateInfo}
+ */
+function makeTemplateInfo({ line, column }, template, isEmbedded) {
+  return {
+    line,
+    column,
+    template,
+    isEmbedded,
+  };
+}
+
+export function coordinatesOf(text, offset) {
+  const contentBeforeTemplateStart = text.slice(0, offset).split('\n');
+  return {
+    line: contentBeforeTemplateStart.length,
+    column: contentBeforeTemplateStart[contentBeforeTemplateStart.length - 1].length,
+  };
+}
+
+/**
+ * @param {TemplateInfo} templateInfo
+ * @param {object} result
+ * @param {number} result.line
+ * @param {number} result.endLine
+ * @param {number} result.column
+ * @param {number} result.endColumn
+ *
+ * @returns {object}
+ */
+export function coordinatesOfResult(templateInfo, result) {
+  /**
+   * Given the sample source code:
+   * 1 import { hbs } from 'ember-cli-htmlbars'\n
+   * 2 export class SomeComponent extends Component<Args> {\n
+   * 3     <template>\n
+   * 4         {{debugger}}\n
+   * 5     </template>\n
+   * 6 }
+   *
+   * The extracted template will be:
+   * 1 \n
+   * 2    {{debugger}}\n
+   *
+   * The coordinates of the template in the source file are: { line: 3, column: 14 }.
+   * The coordinates of the error in the template are: { line: 2, column: 4 }.
+   *
+   * Thus, we need to always subtract one before adding the template location.
```

Comment:
Is this actually true? I think it's only because you are thinking of the newline as not part of the template.

I would expect this to actually not need "line" adjustment, right? (read this as if it were the only contents in a `.gjs` file, which is a valid contents AFAICT)

```js
<template>{{debugger}}</template>
```

Reply:
My understanding is that we want to output the line in the source file, not just the template section.
In the example below, we expect the location of the error to be `{ line: 4, column: 4 }`.
```
   * 1 import { hbs } from 'ember-cli-htmlbars'\n
   * 2 export class SomeComponent extends Component<Args> {\n
   * 3     <template>\n
   * 4         {{debugger}}\n
   * 5     </template>\n
   * 6 }
```
Now the location of the template content is `{ line: 3, column: 14 }` which is where the location after the last character in `<template>` (which is the `\n`).
This means the template is:
```
   * 1 \n
   * 2    {{debugger}}\n
```
The linter will report an error on location `{ line: 2, column: 4 }` of the template. If we simply compute the location in the overall source file without any adjustment the error location ends up being `{ line: 5, column: 4 }` which is incorrect. Thoughts?

Reply:
Gotcha, thank you for the detailed walkthrough (the examples comparing the overall contents of the file wiht line numbers really were helpful).

Reply:
Can you confirm that we have tests for `<template>foo</template>` (errors that happen on the same line as the `<template>` tag)?

Reply:
I had "for a template which does not start with a new line, the error location is correct" already, I also added "for a template defined on one line, the error location is correct" as well.